File: apps/android/app/src/main/java/com/coldtap/hce/data/Session.kt
package com.coldtap.hce.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class SessionStatus {
    CREATED,
    AWAITING_BUYER,
    AWAITING_SIGNATURE,
    SUBMITTED,
    VALIDATING,
    PAID,
    FAILED,
    EXPIRED,
    ;

    val isTerminal: Boolean
        get() = this == PAID || this == FAILED || this == EXPIRED
}

@Serializable
enum class NetworkId {
    @SerialName("mock") MOCK,
    @SerialName("testnet") TESTNET,
    @SerialName("mainnet") MAINNET,
}

@Serializable
data class Session(
    val id: String,
    val merchantName: String,
    val itemName: String,
    val amountDrops: String,
    val amountDisplay: String,
    val currency: String = "XRP",
    val destinationAddress: String,
    val memo: String? = null,
    val status: SessionStatus,
    val txHash: String? = null,
    val network: NetworkId,
    val failureReason: String? = null,
    val expiresAt: String,
    val createdAt: String,
    val updatedAt: String,
    val paidAt: String? = null,
    val failedAt: String? = null,
    val expiredAt: String? = null,
)

@Serializable
data class CreateSessionRequest(
    val merchantName: String,
    val itemName: String,
    val amountDrops: String,
    val destinationAddress: String,
    val memo: String? = null,
    val expiresInSec: Int? = null,
)

@Serializable
data class ApiError(
    val error: String,
    val issues: List<ValidationIssue>? = null,
)

@Serializable
data class ValidationIssue(
    val code: String? = null,
    val message: String,
    val path: List<String>? = null,
)

File: packages/shared/src/session.ts
/**
 * ColdTap canonical session contract.
 *
 * This file is the source of truth for the Session object and status enum.
 * Both the merchant web app and the iOS buyer app mirror these exact shapes.
 * Do not drift.
 */

export const SESSION_STATUSES = [
  "CREATED",
  "AWAITING_BUYER",
  "AWAITING_SIGNATURE",
  "SUBMITTED",
  "VALIDATING",
  "PAID",
  "FAILED",
  "EXPIRED",
] as const;

export type SessionStatus = (typeof SESSION_STATUSES)[number];

export interface Session {
  id: string;
  merchantName: string;
  itemName: string;
  /** Integer-valued string, denominated in XRP drops (1 XRP = 1_000_000 drops). */
  amountDrops: string;
  /** Human-readable amount, e.g. "2.50 XRP". */
  amountDisplay: string;
  currency: "XRP";
  /** Destination XRPL classic address (r-address). */
  destinationAddress: string;
  memo?: string;
  status: SessionStatus;
  /** Final validated XRPL transaction hash. Set once status reaches SUBMITTED or later. */
  txHash?: string;
  expiresAt: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateSessionRequest {
  merchantName: string;
  itemName: string;
  amountDrops: string;
  destinationAddress: string;
  memo?: string;
  /** Optional expiry window in seconds from creation. Defaults to 600 (10 minutes). */
  expiresInSec?: number;
}

export interface SubmitSessionRequest {
  /** Hex-encoded signed XRPL transaction blob from the hardware wallet. */
  txBlob: string;
}

export interface SubmitSessionResponse {
  txHash: string;
  status: Extract<SessionStatus, "SUBMITTED">;
}

export interface StatusUpdateRequest {
  status: SessionStatus;
  txHash?: string;
}
